### PR TITLE
Use AWS SQS Queue system to process tiles

### PR DESCRIPTION
**Pull request by gjn**
This PR will allow us toe use [AWS SQS Queue system](https://aws.amazon.com/sqs/?nc1=h_ls) to handle tiles creation in a distributed manner over several instances. It assures that different instances don't create the same tiles again and again, even when using the exact same configuration.

Boto introduction(http://boto.readthedocs.org/en/latest/sqs_tut.html) and reference(http://boto.readthedocs.org/en/latest/ref/sqs.html)

This adds 4 new make targets:
* `tmscreatequeue`: creates an AWS SQS queue with name specifed in tms configuration which will contain all to-be-created tile coordinates (XYZ) based on the current tms configuration. This is done once usually per dataset that we want to create. Note: each entry consists of maxChunks tiles to be created.
* `tmsqueuestats`: displays some information about an AWS SQS queue, like number of entries.
* `tmsdeletequeue`: will delete a given queue (note: everything is lost!)
* `tmscreatetiles`: will create tiles using the given AWS SQS queue. Do this as many times as you want on as many instances that you want. It's multiprocessing and uses all processes of an instance. It's executed as long as there are tiles/messages in the AWS SQS queue.

Nice to have before running tile creation
[ ] Use db cluseter behind elb instead of direct access. This will allow us to extend the cluster even while running tiles

Proposal:
In order to assure that we always use the same configuration when launching the processing, we should always checkout the same git commit for a given processing batch on all instances.

Note: the creation of the queue can be single processing. on mf0-dev, it took 19 seconds to create messages for 114k tiles. For 20 million tiles, it will take 45 minutes. Also, we can start tile creation as soon as we have started queue creation.

### CODE REVIEW

**Comment by gjn** [comment]
@loicgasser @ltclm 

I think this is ready to be used. It should help us coordinating the processing between different machines. All instances should use the exact same configuration; there's not need to split jobs by hand.

This can still be improved. For instance, when an SQS message can't be processed, we don't handle it (simply ignored). Also, if a tile can't be created, we are simply ignoring it. Both of these things could be written in a re-try SQS queue if we like.

File: forge/lib/tiler.py [hunk 1]
```diff
@@ -245,6 +287,111 @@ def create(self):
...
+        self.t0 = time.time()
+        if len(queueName) <= 0:
+            logger.error('Missing queueName')
+            return
+        try:
+            sqs = getSQS()
+            # queue with default message visiblity time of 600. So each message is blocked
+            # for other users for 600 seconds. Default would be 30 seconds. It's that high
+            # because each message contains maxChunks tiles
+            q = sqs.create_queue(queueName, visibility_timeout=600)
```
Comment (ltclm) [inline forge/lib/tiler.py]: What happens if the queue already exists?

Comment (gjn) [inline forge/lib/tiler.py]: Here, nothing happens. The queue is just returned.

But below (line 311) I check if there is something already in the queue. If yes, it's not created. So we assure that we don't overwrite an existing queue.

File: forge/lib/boto_conn.py [hunk 2]
```diff
@@ -54,6 +55,12 @@ def writeToS3(b, path, content, origin, contentType='application/octet-stream'):
     headers['Content-Encoding'] = 'gzip'
     k.set_contents_from_file(content, headers=headers)
 
+def writeLayerJson(b, filename):
+    headers = {'Content-Type': 'application/json'}
+    k = Key(b)
+    k.key = basePath + 'layer.json'
+    k.set_contents_from_filename(filename, headers=headers)
```
Comment (loicgasser) [inline forge/lib/boto_conn.py]: You could compress it to gzip

File: forge/lib/tiler.py [hunk 3]
```diff
@@ -54,6 +54,50 @@ def grid(bounds, zoomLevels, fullonly):
...
+                    break
+                body = m.get_body()
+                tiles = map(int, body.split(','))
+            except Exception as e:
+                    parseOk = False
+
+            if not parseOk or len(tiles) % 3 != 0:
+                logger.warning('[%s] Unparsable message received. Skipping...and removing message [%s]' % (pid, m.get_body()))
+                q.delete_message(m)
+                continue
```
Comment (loicgasser) [inline forge/lib/tiler.py]: Why is continue needed here?

Comment (gjn) [inline forge/lib/tiler.py]: To get the next message. The while loop should be done until there are no more messages in the sqs queueu. createTimeFromQueue is called only once per process.

File: forge/lib/tiler.py [hunk 4]
```diff
@@ -245,6 +302,118 @@ def create(self):
...
+        nbTiles = self.numOfTiles()
+        try:
+            logger.info('Starting creation of SQS queue with approx. %s tiles)' % (nbTiles))
+            totalcount = 0
+            tcount = 0
+            messagecount = 0
+            msg = ''
+            for tile in tiles:
+                (bounds, tileXYZ, t0, dbConfigFile) = tile
+                if len(msg) != 0:
```
Comment (loicgasser) [inline forge/lib/tiler.py]: Minor: empty string is evaluted to False, so simply `if not msg:`

File: forge/lib/tiler.py [hunk 5]
```diff
@@ -245,6 +302,118 @@ def create(self):
...
+                    msg += ','
+                msg += ('%s,%s,%s' % (str(tileXYZ[0]), str(tileXYZ[1]), str(tileXYZ[2])))
+                tcount += 1
+                if tcount >= maxChunks:
+                    messagecount += 1
+                    writeSQSMessage(q, msg)
+                    tcount = 0
+                    msg = ''
+                totalcount = totalcount + 1
+            if len(msg) > 0:
```
Comment (loicgasser) [inline forge/lib/tiler.py]: same here

File: forge/lib/tiler.py [hunk 6]
```diff
@@ -245,6 +302,118 @@ def create(self):
...
+                messagecount += 1
+                writeSQSMessage(q, msg)
+                pass
```
Comment (loicgasser) [inline forge/lib/tiler.py]: pass is not needed

File: forge/lib/tiler.py [hunk 7]
```diff
@@ -245,6 +302,118 @@ def create(self):
...
+                msg += ('%s,%s,%s' % (str(tileXYZ[0]), str(tileXYZ[1]), str(tileXYZ[2])))
+                tcount += 1
+                if tcount >= maxChunks:
+                    messagecount += 1
+                    writeSQSMessage(q, msg)
+                    tcount = 0
+                    msg = ''
+                totalcount = totalcount + 1
+            if len(msg) > 0:
+                messagecount += 1
```
Comment (loicgasser) [inline forge/lib/tiler.py]: no need to count messages because the number of messages is ` int(math.ceil(float(nbTiles)/float(maxChunks)))`

File: forge/lib/tiler.py [hunk 8]
```diff
@@ -245,6 +302,118 @@ def create(self):
...
+                if len(msg) != 0:
+                    msg += ','
+                msg += ('%s,%s,%s' % (str(tileXYZ[0]), str(tileXYZ[1]), str(tileXYZ[2])))
+                tcount += 1
+                if tcount >= maxChunks:
+                    messagecount += 1
+                    writeSQSMessage(q, msg)
+                    tcount = 0
+                    msg = ''
+                totalcount = totalcount + 1
```
Comment (loicgasser) [inline forge/lib/tiler.py]: Why is this needed?

File: forge/lib/tiler.py [hunk 9]
```diff
@@ -245,6 +302,118 @@ def create(self):
...
+            if len(msg) > 0:
+                messagecount += 1
+                writeSQSMessage(q, msg)
+                pass
+
+        except Exception as e:
+            logger.error('Error during writing of sqs message' + str(e))
```
Comment (loicgasser) [inline forge/lib/tiler.py]: I would raise an expection and eventually delete the queue

File: forge/lib/tiler.py [hunk 10]
```diff
@@ -245,6 +302,118 @@ def create(self):
...
+        tilecount.value = 0
+        skipcount.value = 0
+        queueName = self.tmsConfig.get('General', 'sqsqueue')
+        self.t0 = time.time()
+        if len(queueName) <= 0:
+            logger.error('Missing queueName')
+            return
+        procfactor = int(self.tmsConfig.get('General', 'procfactor'))
+
+        pm = PoolManager(logger=logger, factor = procfactor)
```
Comment (loicgasser) [inline forge/lib/tiler.py]: style not coehrent

File: forge/lib/tiler.py [hunk 11]
```diff
@@ -245,6 +302,118 @@ def create(self):
...
+        queueName = self.tmsConfig.get('General', 'sqsqueue')
+        if len(queueName) <= 0:
+            logger.error('Missing queueName')
+            return
+        try:
+            sqs = getSQS()
+            q = sqs.get_queue(queueName)
+            attrs = sqs.get_queue_attributes(q)
+        except Exception as e:
+            logger.error('Error during deletion of queue' + str(e))
```
Comment (loicgasser) [inline forge/lib/tiler.py]: Wrong message

File: forge/lib/tiler.py [hunk 12]
```diff
@@ -54,6 +54,50 @@ def grid(bounds, zoomLevels, fullonly):
...
+        while True:
+            parseOk = True
+            try:
+                # 20 is maximum wait time
+                m = q.read(visibility_timeout = visibility_timeout, wait_time_seconds = 20)
+                if m is None:
+                    logger.info('[%s] No more messages found. Closing process' % pid)
+                    break
+                body = m.get_body()
+                tiles = map(int, body.split(','))
```
Comment (loicgasser) [inline forge/lib/tiler.py]: try block should be lower if you want to test the parsing only. Or use `ValueError` to test string conversion to int.

File: forge/lib/tiler.py [hunk 13]
```diff
@@ -54,6 +54,50 @@ def grid(bounds, zoomLevels, fullonly):
 tilecount = multiprocessing.Value('i', 0)
 skipcount = multiprocessing.Value('i', 0)
 
+visibility_timeout = 3600
```
Comment (loicgasser) [inline forge/lib/tiler.py]: config? not sure it's better but could make sense

File: .gitignore [hunk 14]
```diff
@@ -10,3 +10,4 @@ logging.cfg
 .*.swo
 apache/testapp.conf
 tmp/*
+forge_*.log
```
Comment (loicgasser) [inline .gitignore]: What is that?

**Comment by loicgasser** [comment]
Thanks
Ok done reviewing, cool usage of AWS queues. Linter to happy please check it out.

File: forge/lib/tiler.py (hunk 3 above)
Comment (loicgasser) [inline forge/lib/tiler.py]: It should continue anyway no?

File: .gitignore (hunk 14 above)
Comment (gjn) [inline .gitignore]: if you don't specify the log directory as env variable, logs are created in root...this avoids adding them to git.

File: forge/lib/tiler.py (hunk 12 above)
Comment (gjn) [inline forge/lib/tiler.py]: maybe name of variable is wrong (parseOk). But this try also checks sqs access and message retrieval, so I would like to keep it like that.

File: forge/lib/tiler.py (hunk 3 above)
Comment (gjn) [inline forge/lib/tiler.py]: we don't know state of `tiles` beyond this point as it depends on where the exception occured when parseOk is False. It's to avoid the for loop, basically.

File: forge/lib/tiler.py (hunk 4 above)
Comment (gjn) [inline forge/lib/tiler.py]: will adapt.

File: forge/lib/tiler.py (hunk 8 above)
Comment (gjn) [inline forge/lib/tiler.py]: Just to have output at the end. `tiles` is lazy evaluated so we can't use len(tiles).

File: forge/lib/tiler.py (hunk 7 above)
Comment (gjn) [inline forge/lib/tiler.py]: I assume that nbTiles is approximately only...

File: forge/lib/tiler.py (hunk 6 above)
Comment (gjn) [inline forge/lib/tiler.py]: indeed

File: forge/lib/tiler.py (hunk 9 above)
Comment (gjn) [inline forge/lib/tiler.py]: It's a bit harsh to delete whole queue just because a simple message could not be written. I agree error hanlding might be better though...

File: forge/lib/tiler.py (hunk 11 above)
Comment (gjn) [inline forge/lib/tiler.py]: copy-paste monster

File: forge/lib/tiler.py (hunk 7 above)
Comment (loicgasser) [inline forge/lib/tiler.py]: ok then

File: .gitignore (hunk 14 above)
Comment (loicgasser) [inline .gitignore]: k

**Comment by loicgasser** [comment]
I'll take care of the corrections as I needed to correct the layer.json

File: forge/lib/tiler.py [hunk 15]
```diff
@@ -245,6 +302,118 @@ def create(self):
...
+            logger.error('Missing queueName')
+            return
+        try:
+            sqs = getSQS()
+            q = sqs.get_queue(queueName)
+            attrs = sqs.get_queue_attributes(q)
+        except Exception as e:
+            logger.error('Error during deletion of queue' + str(e))
+            return
+        print attrs
```
Comment (loicgasser) [inline forge/lib/tiler.py]: logger.info
